pkg/kv/kvserver/store_create_replica.go

```diff
@@ -252,12 +252,12 @@ func fromReplicaIsTooOldRLocked(toReplica *Replica, fromReplica *roachpb.Replica
 	return !found && fromReplica.ReplicaID < desc.NextReplicaID
 }
 
-// addToReplicasByKeyLockedReplicaRLocked adds the replica to the replicasByKey
-// btree. The replica must already be in replicasByRangeID. Returns an error if
-// a different replica with the same range ID, or an overlapping replica or
-// placeholder exists in this Store. Replica.mu must be at least read-locked.
-func (s *Store) addToReplicasByKeyLockedReplicaRLocked(repl *Replica) error {
-	desc := repl.descRLocked()
+// addToReplicasByKeyLocked adds the replica to the replicasByKey btree. The
+// replica must already be in replicasByRangeID. Returns an error if a different
+// replica with the same range ID, or an overlapping replica or placeholder
+// exists in this Store.
```

Comment:
> `repl.mu` must be unlocked

You can sort of "assert" that it is by adding `repl.mu.Lock(); repl.mu.Unlock()` at the top of this method.

Reply:
Or maybe we can push this into `addToReplicasByKeyLocked`, which has multiple callers. Do none of them hold a lock on the repl?

Reply:
There is one caller who holds a lock (`applySnapshot`). This caused a deadlock in this PR.

Reply:
Fixed.